File: src/dr_stone/worker.py
from __future__ import annotations

import argparse
import time

from dr_stone.config import Settings
from dr_stone.logging import configure_logging
from dr_stone.runtime import build_collection_service, build_postgres_storage


DEFAULT_RUNS_PER_DAY = 4
DEFAULT_INTERVAL_SECONDS = int(86400 / DEFAULT_RUNS_PER_DAY)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Run the Dr. Stone scheduled collection worker.")
    parser.add_argument(
        "--run-once",
        action="store_true",
        help="Run one collection cycle immediately and exit.",
    )
    parser.add_argument(
        "--interval-seconds",
        type=int,
        default=DEFAULT_INTERVAL_SECONDS,
        help="Seconds between collection cycles. Defaults to 21600 (4 times per day).",
    )
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    if args.interval_seconds <= 0:
        raise SystemExit("interval-seconds must be a positive integer")

    settings = Settings.from_env()
    logger = configure_logging(settings.log_level)
    storage = build_postgres_storage(logger)
    service = build_collection_service(settings, logger, storage)

    try:
        run_worker_loop(
            service=service,
            logger=logger,
            interval_seconds=args.interval_seconds,
            run_once=args.run_once,
        )
    finally:
        service.close()
    return 0


def run_worker_loop(
    *,
    service,
    logger,
    interval_seconds: int,
    run_once: bool = False,
    sleep_func=time.sleep,
    monotonic_func=time.monotonic,
) -> None:
    while True:
        cycle_started = monotonic_func()
        results = service.collect_all_active()
        logger.info(
            "worker_cycle_completed",
            extra={
                "event_data": {
                    "interval_seconds": interval_seconds,
                    "collected_count": len(results),
                }
            },
        )

        if run_once:
            return

        elapsed_seconds = monotonic_func() - cycle_started
        sleep_seconds = max(0.0, interval_seconds - elapsed_seconds)
        logger.info(
            "worker_sleep_scheduled",
            extra={
                "event_data": {
                    "interval_seconds": interval_seconds,
                    "sleep_seconds": sleep_seconds,
                }
            },
        )
        sleep_func(sleep_seconds)

File: dr-stone-scrapper/src/worker.ts
import process from "node:process";
import { setTimeout as sleep } from "node:timers/promises";

import { loadScrapperSettings } from "./env.js";
import { createLogger } from "./logger.js";
import { buildSearchSources, buildDatabaseServices } from "./runtime.js";
import { AmazonJobScheduler } from "./services/amazon-job-scheduler.js";
import { SearchCollectionService } from "./services/search-collection-service.js";
import type { LoggerLike } from "./types.js";

export async function runWorkerLoop(input: {
  collector: {
    collectAllActive(): Promise<unknown[]>;
  };
  logger: LoggerLike;
  intervalSeconds: number;
  runOnce?: boolean;
  sleepFn?: (milliseconds: number) => Promise<void>;
  now?: () => number;
  scheduleDeferredWork?: () => Promise<{
    scheduledCount: number;
    skippedCount: number;
  }>;
}): Promise<void> {
  const sleepFn = input.sleepFn ?? sleep;
  const now = input.now ?? Date.now;

  while (true) {
    const cycleStarted = now();
    const scheduled =
      (await input.scheduleDeferredWork?.()) ?? {
        scheduledCount: 0,
        skippedCount: 0
      };
    const results = await input.collector.collectAllActive();
    input.logger.info(
      {
        event: "worker_cycle_completed",
        intervalSeconds: input.intervalSeconds,
        collectedCount: results.length,
        scheduledAmazonCount: scheduled.scheduledCount,
        skippedAmazonCount: scheduled.skippedCount
      },
      "worker_cycle_completed"
    );

    if (input.runOnce) {
      return;
    }

    const elapsedMilliseconds = now() - cycleStarted;
    const sleepMilliseconds = Math.max(0, input.intervalSeconds * 1000 - elapsedMilliseconds);
    input.logger.info(
      {
        event: "worker_sleep_scheduled",
        intervalSeconds: input.intervalSeconds,
        sleepSeconds: sleepMilliseconds / 1000
      },
      "worker_sleep_scheduled"
    );
    await sleepFn(sleepMilliseconds);
  }
}

function parseArgs(argv: string[]) {
  const runOnce = argv.includes("--run-once");
  const intervalFlagIndex = argv.indexOf("--interval-seconds");
  const intervalSeconds =
    intervalFlagIndex >= 0 ? Number(argv[intervalFlagIndex + 1]) : undefined;

  return {
    runOnce,
    intervalSeconds
  };
}

export async function main(argv = process.argv.slice(2)): Promise<number> {
  await import("dotenv/config");
  const args = parseArgs(argv);
  if (args.intervalSeconds !== undefined && args.intervalSeconds <= 0) {
    throw new Error("interval-seconds must be a positive integer");
  }

  const configuredSources = process.env.DR_STONE_ENABLED_SOURCES;
  const settings = loadScrapperSettings({
    intervalSeconds: args.intervalSeconds
  });

  if (settings.intervalSeconds <= 0) {
    throw new Error("interval-seconds must be a positive integer");
  }

  const logger = createLogger(settings.logLevel);
  logger.info(
    {
      event: "worker_started",
      runOnce: args.runOnce,
      intervalSeconds: settings.intervalSeconds,
      enabledSources: settings.enabledSources,
      sourceConfiguration: configuredSources ? "environment" : "default"
    },
    "worker_started"
  );

  if (!configuredSources) {
    logger.warn(
      {
        event: "worker_default_sources_used",
        defaultEnabledSources: settings.enabledSources,
        environmentVariable: "DR_STONE_ENABLED_SOURCES"
      },
      "worker_default_sources_used"
    );
  }

  const database = await buildDatabaseServices(settings);
  const searchSources = buildSearchSources(settings, logger);
  const immediateSources = searchSources.filter((source) => source.sourceName !== "amazon");
  const amazonSources = searchSources.filter((source) => source.sourceName === "amazon");
  const collector =
    immediateSources.length > 0
      ? new SearchCollectionService(database, immediateSources, logger)
      : {
          collectAllActive: async () => []
        };
  const amazonService =
    amazonSources.length > 0
      ? new SearchCollectionService(database, amazonSources, logger)
      : null;
  const amazonScheduler =
    amazonService && amazonSources.length > 0
      ? new AmazonJobScheduler(settings, database, amazonService, logger)
      : null;

  logger.info(
    {
      event: "worker_sources_partitioned",
      immediateSources: immediateSources.map((source) => source.sourceName),
      deferredSources: amazonSources.map((source) => source.sourceName),
      amazonMinIntervalSeconds: settings.amazonMinIntervalSeconds
    },
    "worker_sources_partitioned"
  );

  try {
    if (amazonScheduler) {
      await amazonScheduler.start();
    }

    await runWorkerLoop({
      collector,
      logger,
      intervalSeconds: settings.intervalSeconds,
      runOnce: args.runOnce,
      scheduleDeferredWork: amazonScheduler
        ? () => amazonScheduler.scheduleActiveTrackedProducts()
        : undefined
    });
  } finally {
    await amazonScheduler?.stop();
    if (collector instanceof SearchCollectionService) {
      await collector.close();
    }
    await amazonService?.close();
    await database.close();
  }

  return 0;
}

if (process.argv[1] && process.argv[1].endsWith("worker.js")) {
  main().catch((error: unknown) => {
    process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
    process.exitCode = 1;
  });
}
